### ISSUE TITLE
cdk proposal: framework for rich text editors

### ISSUE THREAD

**Issue by fxck** [angular/material2#4444]
tl;dr https://draftjs.org/ for angular

I've been searching the internets and there is nothing that comes even close to draftjs, which is built on top and requires react. What does it do? It solves most of the [contenteditable issues](https://medium.engineering/why-contenteditable-is-terrible-122d8a40e480) and provides api for people to build their own rich text editors on top of it.

Rich text editor doesn't necessarily mean WYSIWYG textarea, it can help with anything from mentions, hashtags to replacing text emoticons with images etc.

![image](https://cloud.githubusercontent.com/assets/1303561/25852521/edeec4ea-34c9-11e7-928c-394a8ffa8ed5.png)

Some of the most popular draftjs plugins can be found [here](https://www.draft-js-plugins.com/) and [here](https://github.com/nikgraf/awesome-draft-js), it should give you an idea what can be done with it.

worth reading:
- https://code.facebook.com/posts/1684092755205505/facebook-open-sources-rich-text-editor-framework-draft-js/
- https://medium.com/@rajaraodv/why-draft-js-and-why-you-should-contribute-460c4a69e6c8

I think this could be a great fit for CDK, given that you are eventually planning on creating WYSIWYG anyway.. what do you think?

@jelbourn @mmalerba @kara @andrewseguin @DevVersion @crisbeto

**Comment by DevVersion** [comment]
Thanks for the links. Why aren't we moving that conversation to the issue https://github.com/angular/material2/issues/2949?

**Comment by fxck** [comment]
Because that guy's looking for WYSIWYG, this is something you'd build WYSIWYG on top of, two very different things I think.

**Comment by DevVersion** [comment]
Not in my opinion. As you said, the editor would have been built on top of that. So it relates to that conversation.

It's like you a separate issue just for the menu-item of a menu.

**Comment by fxck** [comment]
Nope, it's like a different issue for Overlay and dialog/menu/select.

**Comment by DevVersion** [comment]
The issue is pretty generic and can be used for a "Rich Text Editor" too. Anyways we'll see if something like this should be part of the CDK at all.

My personal opinion is that something like that would be rather part of another repository or just a 3rd party library.

**Comment by fxck** [comment]
I understand why you'd think so, but look into what draftjs actually does (or watch this video https://www.youtube.com/watch?v=feUYwoLhE_4), it basically doesn't care about visual representation at all, which what that guy in #2949 is ultimately looking for a component with buttons that do stuff..

The reason I'm really asking here is because google/angular team has enough brain power and experience to actually do something like this. If you search the internet you won't find anything remotely close to draftjs, everyone's creating mentions inputs, emoticon replacers, WYSIWYG textareas, only facebook took a step back and asked what's in the core of such features. I really find it similar to what other CDK services do. It could greatly empower other component creators, it's a really powerful high level service.

**Comment by mlc-mlapis** [comment]
I did not go to the details but it looks that https://quilljs.com (used in PrimeNG UI components for Angular https://www.primefaces.org/primeng/#/editor) is very similar from the concept point of view?

**Comment by fxck** [comment]
Quill looks like WYSIWYG to me. If there's anything similar, it's [slate](https://github.com/ianstormtaylor/slate), again, just for react. Read the [comparisons](https://github.com/ianstormtaylor/slate/blob/master/docs/general/comparisons.md).

Also https://youtu.be/gxNuHZXZMgs?t=1791 this guy saying that he talked to angular people who decided to add react/draft as a dependency simply because they couldn't find anything similar in angular world.

**Comment by mlc-mlapis** [comment]
From the first point of view sure but I meant mainly the background of it. Its internal structure, API, ...

**Comment by fxck** [comment]
I don't know, I think their own comparison to draft speaks for itself https://quilljs.com/guides/comparison-with-other-rich-text-editors/#draft they are out of the box usable wysiwyg, draft is just a high level service that allows you to build rich text editors (not necessarily wysiwyg editors).

Simple test is to see if quill, or any other lib, can also be used to smaller, simple things like facebook-like mentions. You won't really find a quill plugin for that, and if you do, you'll find that it relies on DOM data-* attributes to carry data, unlike draft which stores them in a JSON structure, which I think is a better approach in angular world.

**Comment by mlc-mlapis** [comment]
Yeah, it is certainly true. I did not go so deep, I just looked at some basic operation as
```
quill.setContents([
  { insert: 'Hello ' },
  { insert: 'World!', attributes: { bold: true } },
  { insert: '\n' }
]);
```
and corresponding `Delta`s representing the change which use JSON format.

**Comment by jelbourn** [comment]
We have chatted about this in the past (and one Angular team member actually has a decent amount of experience with this kind of thing). It's something we'd like to do, but is pretty far down on the priority list.

**Comment by mlc-mlapis** [comment]
Oh, really? Sounds positive. 👍 Is it possible to share some more details?

**Comment by fxck** [comment]
@jelbourn I got in touch with the guy who made draftjs, to see if he or the team would be willing to share experience and help you out, he's not working at facebook anymore but said he'd pass the message along.

**Comment by fxck** [comment]
So I started messing with it yesterday, I managed to port the whole model part to typescript, you can check the prototype [here](https://github.com/thescrollbar/angular-draft).

With it, you can create [EditorState](https://draftjs.org/docs/api-reference-editor-state.html#content), [ContentState](https://draftjs.org/docs/api-reference-content-state.html#content), [SelectionState](https://draftjs.org/docs/api-reference-selection-state.html#content), which is pretty cool, now there are two more major parts missing, first is the actual fix of contentEditable, which I'm gonna work on today and which should be possible to more of less copy from draft, the other part are [decorators](https://draftjs.org/docs/advanced-topics-decorators.html#content), I don't think it's even possible to do something like that in angular, so I'm gonna look for some help.

**Comment by axtho** [comment]
@fxck are you going to continue with the prototype? The suggestion is really good.

**Comment by fxck** [comment]
@axtho I ran out of time I could dedicate to it, tried to get couple of people on board, but none was really interested

**Comment by fxck** [comment]
@jelbourn now that almost all of the components are done (or at least will be by the end of the year), any approximate idea when this could happen? Q1 / Q2 18?

**Comment by jelbourn** [comment]
We might start exploring the space in Q2 depending on how other stuff goes. In terms of larger projects, virtual-scrolling, tree, table improvements, and data visualization are all higher up on the list.

**Comment by fxck** [comment]
The way angular.io handles docs / https://github.com/wardbell/ng-dynamic-app/blob/master/src/app/docviewer/docviewer.component.ts might actually be a way (and probably the only way without including the whole compiler) to do something like draftjs decorators https://draftjs.org/docs/advanced-topics-decorators.html#content

**Comment by fxck** [comment]
@jelbourn Q2 still a go for the exploration?

**Comment by jelbourn** [comment]
Probably not at this point; it's looking now that Q2 is mostly going to be finishing up drag/drop and virtual, addressing some tech debt and productivity issues (mainly trying to speed up the rate at which we can sync stuff into Google), and support/validation for Ivy work.

**Comment by wynfred** [comment]
I wrote a [directive](https://github.com/wynfred/ng-mat-contenteditable/blob/master/projects/mat-contenteditable/src/lib/mat-contenteditable.directive.ts). It implements MatFormFieldControl to make `contenteditable` compatible with MatFormField.
I'm pretty new to web development, so I basically just copy/paste stuff from matInput. The `errorstate`  and some other stuff is a little confusing for me and I'm not sure I implemented them correctly. I hope I can get some help from material team to test/improve/fix my implementation.

**Comment by naveedahmed1** [comment]
@jelbourn are there any guidelines for third-party WYSIWYG editor developers who want to add support for Angular Material 2?

For example CKEditor 5 which has been completely rewritten in ES6 and provides all the necessary tools to easily integrate it with modern applications and technologies such as Angular. And they officially support Angular https://github.com/ckeditor/ckeditor5-angular. They seems to be interested in officially supporting Angular Material as well but clarity on best approach is needed.

Please take a look at:
https://github.com/ckeditor/ckeditor5-angular/issues/21
and
https://github.com/ckeditor/ckeditor5-angular/issues/21#issuecomment-427815650

**Comment by jelbourn** [comment]
It's not clear to me what "adding support for Angular Material" would mean. Could you clarify?

**Comment by naveedahmed1** [comment]
@jelbourn To be more specific I am referring to the support of WYSIWYG editor inside `MatFormFieldControl`. In case of ckeditor they have a separate plugin `ckeditor/ckeditor5-angular` for Angular.

Now, to support ckeditor inside `MatFormFieldControl`, should they have a separate plugin e.g `ckeditor5-angular-material`? since in order to support `MatFormFieldControl` they have will to add `@angular/material` a peer or explicit dependency in `package.json` file. And since its not necessary that everyone using `ckeditor/ckeditor5-angular` would also be using Angular Material, so it would be an additional dependency for them. Or is there any other solution that can allow this without importing it to satisfy the angular DI.

**Comment by jelbourn** [comment]
All you would really need is a directive that implements `MatFormFieldControl` in order to work with `mat-form-field` (similar to `MatInput`). That's something that would best fit in its own package so any existing packages don't have to pull in dependencies.

**Comment by fxck** [comment]
Any updates?

**Comment by wynfred** [comment]
Not sure if the material team is interested in taking something like [this](https://github.com/wynfred/mat-contenteditable/blob/master/projects/mat-contenteditable/src/lib/mat-contenteditable.directive.ts) into the official lib.

**Comment by jelbourn** [comment]
No plans to work on this at least through 2020.

**Comment by fxck** [comment]
https://github.com/facebook/lexical Facebook now created a better and improved draft.js
